src/pipelinedefinition.cpp

```diff
@@ -0,0 +1,322 @@
+
+    EntryNode* entry = nullptr;
+    ExitNode* exit = nullptr;
+    for (const auto& info : nodeInfos) {
+        SPDLOG_DEBUG("Creating pipeline:{}. Adding nodeName:{}, modelName:{}",
+            getName(), info.nodeName, info.modelName);
+        switch (info.kind) {
+        case NodeKind::ENTRY: {
+            auto node = std::make_unique<EntryNode>(request);
+            entry = node.get();
+            nodes.insert(std::make_pair(info.nodeName, std::move(node)));
+            break;
+        }
+        case NodeKind::DL:
+            nodes.insert(std::make_pair(info.nodeName, std::move(std::make_unique<DLNode>(info.nodeName,
+                                                           info.modelName,
+                                                           info.modelVersion,
+                                                           manager,
+                                                           info.outputNameAliases))));
+            break;
+        case NodeKind::EXIT: {
+            auto node = std::make_unique<ExitNode>(response);
+            exit = node.get();
+            nodes.insert(std::make_pair(info.nodeName, std::move(node)));
+            break;
+        }
+        default:
+            throw std::invalid_argument("unknown node kind");
+        }
+    }
+    for (const auto& kv : connections) {
+        const auto& dependantNode = nodes.at(kv.first);
+        for (const auto& pair : kv.second) {
+            const auto& dependencyNode = nodes.at(pair.first);
+            SPDLOG_DEBUG("Connecting pipeline:{}, from:{}, to:{}", getName(), dependencyNode->getName(), dependantNode->getName());
+            Pipeline::connect(*dependencyNode, *dependantNode, pair.second);
+        }
+    }
+    pipeline = std::make_unique<Pipeline>(*entry, *exit, pipelineName);
+    for (auto& kv : nodes) {
+        pipeline->push(std::move(kv.second));
+    }
+    return StatusCode::OK;
+}
+
+void PipelineDefinition::resetSubscriptions(ModelManager& manager) {
+    for (auto& [modelName, modelVersion] : subscriptions) {
+        if (modelVersion) {
+            SPDLOG_INFO("Unsubscribing pipeline:{} from model: {}, version:{}",
+                getName(), modelName, modelVersion);
+            manager.findModelByName(modelName)->getModelInstanceByVersion(modelVersion)->unsubscribe(*this);
+        } else {  // using default version
+            SPDLOG_INFO("Unsubscribing pipeline:{} from model: {}",
+                getName(), modelName);
+            manager.findModelByName(modelName)->unsubscribe(*this);
+        }
+    }
+    subscriptions.clear();
+}
+
+void PipelineDefinition::makeSubscriptions(ModelManager& manager) {
+    for (auto& node : nodeInfos) {
+        if (node.kind == NodeKind::DL) {
+            if (subscriptions.find({node.modelName, node.modelVersion.value_or(0)}) != subscriptions.end()) {
+                continue;
+            }
+            std::unique_ptr<ModelInstanceUnloadGuard> nodeModelInstanceUnloadGuard;
```

Comment:
Why do you need model guard to subscribe but not for unsubscribe?

Reply:
Actually I don't need it here - I used getModelInstance for convenience. I will change to

findModelByName
if(node.modelVersion) {
model->getModelInstanceByVersion(node.modelVersion)->subscribe()
} else {
model->subscribe()
}